pkg/kubelet/status/status_manager.go

```diff
@@ -352,12 +352,22 @@ func (m *manager) SetPodStatus(pod *v1.Pod, status v1.PodStatus) {
 	// Make sure we're caching a deep copy.
 	status = *status.DeepCopy()
 
+	// Set the observedGeneration for this pod status. Static pods are excluded.
+	if podObservedGenerationTrackingEnabled(pod) && !metav1.HasAnnotation(pod.ObjectMeta, v1.MirrorPodAnnotationKey) {
+		status.ObservedGeneration = pod.GetGeneration()
+	}
+
 	// Force a status update if deletion timestamp is set. This is necessary
 	// because if the pod is in the non-running state, the pod worker still
 	// needs to be able to trigger an update and/or deletion.
 	m.updateStatusInternal(pod, status, pod.DeletionTimestamp != nil, false)
 }
 
+// We will emit status.observedGeneration if the feature is enabled OR if status.observedGeneration is already set.
+func podObservedGenerationTrackingEnabled(pod *v1.Pod) bool {
+	return pod.Status.ObservedGeneration != 0 || utilfeature.DefaultFeatureGate.Enabled(features.PodObservedGenerationTracking)
```

Comment:
a note to myself (and early reviewers) before I take this out of draft mode: I need to double check that this works, I'm not sure that the old pod status actually gets propagated all the way here

Reply:
> I'm not sure that the old pod status actually gets propagated all the way here

confirmed that it does and that observedGeneration continues to get updated (if already set) even after the feature is disabled, so we are good here